localstack/extensions/patterns/webapp.py

```diff
@@ -0,0 +1,321 @@
+import importlib
+import logging
+import mimetypes
+import typing as t
+from functools import cached_property
+
+from rolo.gateway import HandlerChain
+from rolo.router import RuleAdapter, WithHost
+from werkzeug.routing import Submount
+
+from localstack import config
+from localstack.aws.api import RequestContext
+from localstack.extensions.api import Extension, http
+
+if t.TYPE_CHECKING:
+    # make sure jinja2 stays optional
+    import jinja2
+
+LOG = logging.getLogger(__name__)
+
+_default = object()
+
+
+class WebAppExtension(Extension):
+    """
+    EXPERIMENTAL! This class is experimental and the API may change without notice.
+
+    A webapp extension serves routes, templates, and static files via a submount and a subdomain through localstack.
+
+    It assumes you have the following directory layout::
+
+        my_extension
+        ├── extension.py
+        ├── __init__.py
+        ├── static              <-- make sure static resources get packaged!
+        │   ├── __init__.py
+        │   ├── favicon.ico
+        │   └── style.css
+        └── templates            <-- jinja2 templates
+            └── index.html
+
+    Given this layout, you can define your extensions in ``my_extension.extension`` like this. Routes defined in the
+    extension itself are automatically registered::
+
+        class MyExtension(WebAppExtension):
+            name = "my-extension"
+
+            @route("/")
+            def index(request: Request) -> Response:
+                # reference `static/style.css` to serve the static file from your package
+                return self.render_template_response("index.html")
+
+            @route("/hello")
+            def hello(request: Request):
+                return {"message": "Hello World!"}
+
+    This will create an extension that localstack serves via:
+
+    * Submount: https://localhost.localstack.cloud:4566/_extension/my-extension
+    * Subdomain: https://my-extension.localhost.localstack.cloud:4566/
+
+    Any routes added by the extension will be served relative to these URLs.
+    """
+
+    def __init__(
+        self,
+        mount: str = None,
+        submount: str | None = _default,
+        subdomain: str | None = _default,
+        template_package_path: str | None = _default,
+        static_package_path: str | None = _default,
+        static_url_path: str = None,
+    ):
+        """
+        Overwrite to customize your extension. For example, you can disable certain behavior by calling ``super(
+        ).__init__(subdomain=None, static_package_path=None)``, which will disable serving through a subdomain,
+        and disable static file serving.
+
+        :param mount: the "mount point" which will be used as default value for the submount and subdirectory, i.e.,
```

Comment:
What happens upon naming clashes? the last registered route wins 🤔

Reply:
True. nothing we can do about that right now unfortunately. Added a comment though